cns/restserver/ipam.go

```diff
@@ -478,101 +634,194 @@ func (service *HTTPRestService) MarkExistingIPsAsPendingRelease(pendingIPIDs []s
+		IPMap[desiredIP] = desiredIP
+	}
+	ncMap := make(map[string]cns.IPConfigurationStatus)
+	assignedMap := make(map[string]cns.IPConfigurationStatus)
+
+	for _, ipConfig := range service.PodIPConfigState { //nolint:gocritic // ignore copy
+		_, found := IPMap[ipConfig.IPAddress]
+		// keep searching until the all desired IPs are found
+		if !found {
+			continue
+		}
+		switch ipConfig.GetState() { //nolint:exhaustive // ignoring PendingRelease case intentionally
+		case types.Assigned:
+			// This IP has already been assigned, if it is assigned to same pod, then return the same
+			// IPconfiguration
+			if ipConfig.PodInfo.Key() == podInfo.Key() {
+				logger.Printf("[AssignDesiredIPConfigs]: IP Config [%+v] is already assigned to this Pod [%+v]", ipConfig, podInfo)
+				assignedMap[ipConfig.NCID] = ipConfig
+				err := service.populateIPConfigInfoUntransacted(ipConfig, &podIPInfo[len(assignedMap)])
+				if len(assignedMap) == len(desiredIPAddresses) {
+					return podIPInfo, err
 				}
-			default:
-				return podIpInfo, errors.Errorf("[AllocateDesiredIPConfig] Desired IP is not available %+v", ipConfig)
+			} else {
+				logger.Errorf("[AssignDesiredIPConfigs] Desired IP is already assigned %+v, requested for pod %+v", ipConfig, podInfo)
+				//nolint:goerr113 // return error
+				return podIPInfo, fmt.Errorf("IP already assigned to another pod")
+			}
+		case types.Available, types.PendingProgramming:
+			// This race can happen during restart, where CNS state is lost and thus we have lost the NC programmed version
+			ncMap[ipConfig.NCID] = ipConfig
+		default:
+			logger.Errorf("[AssignDesiredIPConfigs] Desired IP is not available %+v", ipConfig)
+			//nolint:goerr113 // return error
+			return podIPInfo, fmt.Errorf("IO not available")
+		}
+		if len(ncMap) == len(desiredIPAddresses) {
+			break
+		}
+	}
+
+	podIPctr := 0
+	// assigns all IPs in the map to the pod
+	for id := range ncMap {
+
+		// As part of reconcile, we mark IPs as Assigned which are already assigned to Pods (listed from APIServer)
+		if err := service.assignIPConfig(ncMap[id], podInfo); err != nil {
+			logger.Errorf(err.Error())
+			break
+		}
+
+		if err := service.populateIPConfigInfoUntransacted(ncMap[id], &podIPInfo[podIPctr]); err != nil {
+			logger.Errorf(err.Error())
+			break
+		}
+		podIPctr++
+		// Checks if we no longer have any IPs to be added and returns if we don't
+		if podIPctr == len(service.state.ContainerStatus) {
+			return podIPInfo, nil
+		}
+	}
+
+	// if we were able to get at least one IP but not all of the desired IPs
+	if len(ncMap) > 0 {
+		logger.Printf("[AssignDesiredIPConfigs] Failed to retrieve all desired IPs. Releasing all IPs that were found")
+		for _, ipState := range ncMap { //nolint:gocritic // ignore copy
+			_, err := service.unassignIPConfig(ipState, podInfo)
+			if err != nil {
+				return podIPInfo, fmt.Errorf("[AssignDesiredIPConfigs] failed to mark IPConfig [%+v] back to Available. err: %w", ipState, err)
```

Comment:
this error would not make sense for the caller. Caller called to assign the IPs and it's getting the error saying 'failed to mark IP back to available'. What is it supposed to do in case it gets that error?